perfkitbenchmarker/providers/openstack/os_disk.py

```diff
@@ -24,96 +24,198 @@
+      return volume and volume.status in ('available', 'in-use',
+                                          'deleting',)
+    except NotFound:
+      return False
+
+  def Attach(self, vm):
+    self.attached_vm_name = vm.name
+    self.attached_vm_id = vm.id
+
+    result = self.__nclient.volumes.create_server_volume(vm.id,
+                                                         self._disk.id,)
+    self.attach_id = result.id
+
+    volume = None
+    is_unattached = True
+    while is_unattached:
+      time.sleep(1)
+      volume = self.__nclient.volumes.get(result.id)
+      if volume:
+          is_unattached = not(volume.status == "in-use"
+                              and volume.attachments)
+
+    for attachment in volume.attachments:
+      if self.attach_id == attachment.get('volume_id'):
+        self.device = attachment.get('device')
+        return
+
+    raise errors.Error("Couldn't not attach volume to %s" % vm.name)
+
+  def GetDevicePath(self):
+    return self.device
+
+  def Detach(self):
+    self.__nclient.volumes.delete_server_volume(self.attached_vm_id,
+                                                self.attach_id)
+
+
+class OpenStackLocalDisk(OpenStackDisk):
+  """
+  OpenStackLocalDisk is a disk object to represent ephemeral storage disk that
+  is locally attached to an instance.
+  """
+
+  def __init__(self, disk_spec, device_path, name):
+    """Initialize OpenStack Local Disk object.
+
+    Args:
+      disk_spec: DiskSpec object contains information needed of desired disk.
+      device_path: string Linux device path of the local disk. e.g. '/dev/vdb'
+    """
+    super(OpenStackLocalDisk, self).__init__(disk_spec, name)
+    self.exists = False
+    self.device_path = device_path
+    self.name = 'local-disk-%s' % name
+
+  def _Create(self):
+    self.exists = True
+
+  def _Exists(self):
+    return self.exists
+
+  def _Delete(self):
+    self.exists = False
+
+  def Attach(self, vm):
+    pass
+
+  def Detach(self):
+    pass
+
+
+class OpenStackRootDisk(OpenStackLocalDisk):
+  """
+  OpenStackRootDisk is a disk object to represent the root disk of an instance.
+  Root disk provide a directory path as a scratch disk space for a benchmark,
+  but does not allows its backing block device to be formatted, or its mount
+  point to be changed.
+  """
+
+  def __init__(self, disk_spec, device_path):
+    """Initialize OpenStack Local Disk object.
+
+    Args:
+      disk_spec: DiskSpec object contains information needed of desired disk.
+      device_path: string Linux device path of the local disk. e.g. '/dev/vdb'
+    """
+    super(OpenStackRootDisk, self).__init__(disk_spec, device_path, 'root-disk')
+    self.mount_point = disk_spec.mount_point
+
+  def GetDevicePath(self):
+    """
+    Root disk is already mounted. Formatting of of root disk is not permitted.
+    """
+    raise errors.Error('Formatting of root disk block device is not supported.')
```

Comment:
I don't think we should raise an exception here, because it is possible to get the device path of a boot disk, and we might want to. For instance, a benchmark might `ls -l` the device path to figure out how big the disk is.

It might make more sense to have an `Unmount()` or `FormatDevice()` method, and either of those would raise an exception for a boot disk.